Sources/Operators/RetryWhen.swift

```diff
@@ -0,0 +1,102 @@
+    func retryWhen<Trigger>(_ notificationHandler: @escaping (AnyPublisher<Self.Failure, Never>) -> Trigger)
+        -> Publishers.RetryWhen<Self, Trigger, Output, Failure> where Trigger: Publisher {
+            .init(upstream: self, notificationHandler: notificationHandler)
+    }
+}
+
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+public extension Publishers {
+    class RetryWhen<Upstream, Trigger, Output, Failure>: Publisher where Upstream: Publisher, Upstream.Output == Output, Upstream.Failure == Failure, Trigger: Publisher {
+        
+        typealias Handler = (AnyPublisher<Upstream.Failure, Never>) -> Trigger
+        
+        private let upstream: Upstream
+        private let handler: Handler
+        
+        init(upstream: Upstream, notificationHandler: @escaping Handler) {
+            self.upstream = upstream
+            self.handler = notificationHandler
+        }
+        
+        public func receive<S>(subscriber: S) where S: Subscriber, Failure == S.Failure, Output == S.Input {
+            subscriber.receive(subscription: Subscription(upstream: upstream, downstream: subscriber, handler: handler))
+        }
+    }
+}
+
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+extension Publishers.RetryWhen {
+    class Subscription<Downstream>: Combine.Subscription where Downstream: Subscriber, Downstream.Input == Upstream.Output, Downstream.Failure == Upstream.Failure {
+        
+        private let upstream: Upstream
+        private let downstream: Downstream
+        private let errorSubject = PassthroughSubject<Upstream.Failure, Never>()
+        private var sink: Sink<Upstream, Downstream>?
+        private var cancellable: AnyCancellable?
+        
+        init(upstream: Upstream, downstream: Downstream, handler: @escaping (AnyPublisher<Upstream.Failure, Never>) -> Trigger) {
+            self.upstream = upstream
+            self.downstream = downstream
+            self.sink = Sink(
+				upstream: upstream,
+				downstream: downstream,
+                transformOutput: { $0 },
+                transformFailure: { [errorSubject] in
+                    errorSubject.send($0)
+                    return nil
+                }
+            )
+            self.cancellable = handler(errorSubject.eraseToAnyPublisher())
+                .sink(
+                    receiveCompletion: { [sink] completion in
+                        switch completion {
+                        case .finished:
+                            sink?.buffer.complete(completion: Subscribers.Completion<Downstream.Failure>.finished)
+                        case .failure(let error):
+                            if let error = error as? Downstream.Failure {
+                                sink?.buffer.complete(completion: Subscribers.Completion<Downstream.Failure>.failure(error))
+                            }
```

Comment:
```suggestion
                            sink?.buffer.complete(completion: >.finished)
                        case .failure(let error):
                            if let error = error as? Downstream.Failure {
                                sink?.buffer.complete(completion: .failure(error))
                            }
```

I think inference should work here?